README
#使用方法

#安装mentohust：建议Ubuntu用户使用Deb包安装，Fedora用户使用RPM包安装

#如果确定自己可以使用xrgsu认证成功，打开终端输入sudo mentohust运行即可
#如果不确定，切换到锐捷所在目录，然后输入以下命令：
sudo mkdir /etc/mentohust
sudo cp ./8021x.exe  /etc/mentohust
#如果准确按以上步骤操作后还是认证失败，请下载MentoHUSTTool，在Windows下抓包并保存为data.mpf，
然后回到Linux，切换到data.mpf所在目录，输入以下命令：
sudo cp ./data.mpf /etc/mentohust
然后打开终端输入sudo mentohust -f/etc/mentohust/data.mpf -w运行即可。以后也只需输入sudo mentohust。

#如何退出:不以后台模式运行mentohust时，按Ctrl+C或者关闭终端即可退出；后台运行时使用sudo mentohust -k退出认证。

#查看帮助信息请输入：mentohust -h
更详细的帮助信息请参考：http://wiki.ubuntu.org.cn/锐捷、赛尔认证MentoHUST

#修改参数请根据帮助信息操作，例如修改用户名和密码：sudo mentohust -uUsername -pPassword -w
指定某些参数仅对当次认证有效请根据帮助信息操作，，例如临时修改用户名和密码：sudo mentohust -uUsername -pPassword

#如果提示缺少libpcap.so.0.x而在/usr/lib/目录下已存在一个libpcap.so.0.x.y，输入以下命令：
sudo ln -s /usr/lib/libpcap.so.0.x.y /usr/lib/libpcap.so.0.x
否则请安装libpcap。

#权责声明
1、本程序所有涉及锐捷赛尔认证的功能均是来自前辈公开代码及抓包分析。
2、本程序于个人仅供学习，于他人仅供方便认证，不得使用本程序有意妨害锐捷赛尔认证机制及相关方利益。
3、一切使用后果由用户自己承担。
4、本程序不提供任何服务及保障，编写及维护纯属个人爱好，随时可能被终止。
5、使用本程序者，即表示同意该声明。谢谢合作。

源码可在项目主页获取：http://mentohust.googlecode.com/
联系作者：在http://mentohust.googlecode.com/留言或Email：mentohust@ehust.co.cc


ChangeLog
2011-02-07 12:10  schemacs@gmil.com

	* src/config.c: 配置参数时输入密码更改为不回显
	* po/en.po: 更正部分英文翻译
	* ChangeLog: 用svn2cl生成ChangeLog历史内容

2011-01-30 23:57  microcaicai

	* mentohust.spec.in: rpm spec fix

2011-01-30 23:56  microcaicai

	* po, po/en.po: Update english translation

2010-10-29 00:44  microcaicai

	* src/mycheck.c: 好的，添加上去

2010-10-27 04:55  www.ehust@gmail.com

	* src/dlfunc.c, src/mentohust.c, src/mycheck.c, src/myfunc.c: 字节对齐

2010-06-02 07:14  robinlee.sysu@gmail.com

	* mentohust.spec.in, src/mycheck.c: update the spec file and
	  include config.h in src/mycheck.c

2010-05-14 12:20  www.ehust@gmail.com

	* po/en.po, src/dlfunc.c, src/dlfunc.h, src/mentohust.c,
	  src/myconfig.c: 解决Ubuntu10.04下libnotify的问题

2010-04-23 02:02  www.ehust@gmail.com

	* po/en.po: 忘了svn add，这个好容易忘啊

2010-04-22 13:38  www.ehust@gmail.com

	* Makefile.am, README, autogen.sh, configure.ac, po/LINGUAS,
	  po/Makevars, po/POTFILES.in, po/zh_CN.po, src/Makefile.am,
	  src/dlfunc.c, src/dlfunc.h, src/i18n.h, src/mentohust.c,
	  src/mentohust.conf, src/mycheck.c, src/mycheck.h, src/myconfig.c,
	  src/myfunc.c, src/myfunc.h, src/myini.c, src/myini.h,
	  src/mystate.c: 新增-q参数显示解密后的SuConfig.dat以便选择DHCP方式。

2010-04-19 08:58  microcaicai

	* src/mentohust.c: 修复错误的预处理

2010-04-09 11:28  microcaicai@gmail.com

	* autogen.sh: change missing w

2010-03-28 21:21  microcaicai

	* mentohust.spec.in: rpm 包 改一下

2010-03-28 21:20  microcaicai

	* Makefile.am, configure.ac, mentohust.spec, mentohust.spec.in: rpm
	  打包机制改革

2010-03-28 21:16  microcaicai

	* ., Makefile.am: 改改。。。。

2010-03-28 21:13  microcaicai

	* po/zh_CN.po, src/myconfig.c: 翻译一下

2010-03-28 21:04  microcaicai

	* po/zh_CN.po, src/dlfunc.c: 翻译一下

2010-03-28 20:49  microcaicai

	* po/zh_CN.po: 翻译一下

2010-03-28 20:43  microcaicai

	* configure.ac: 好看点的编译输出显示

2010-03-28 20:37  microcaicai

	* ., Makefile.am, configure.ac, m4, po, po/LINGUAS, po/Makevars,
	  po/POTFILES.in, po/zh_CN.po, src, src/dlfunc.c, src/dlfunc.h,
	  src/i18n.h, src/mentohust.c, src/mycheck.c, src/myconfig.c,
	  src/myfunc.c, src/myini.c, src/mystate.c: 国际化，这次是真的可以用了

2010-03-10 06:16  www.ehust@gmail.com

	* ., AUTHORS, ChangeLog, Makefile.am, NEWS, README, autogen.sh,
	  configure.ac, debian, debian/changelog, debian/compat,
	  debian/control, debian/copyright, debian/rules, debian/watch,
	  mentohust.1, mentohust.spec, src, src/Makefile.am, src/dlfunc.c,
	  src/dlfunc.h, src/md5.c, src/md5.h, src/mentohust.c,
	  src/mentohust.conf, src/mycheck.c, src/mycheck.h, src/myconfig.c,
	  src/myconfig.h, src/myfunc.c, src/myfunc.h, src/myini.c,
	  src/myini.h, src/mystate.c, src/mystate.h, src/types.h:
	  针对FreeBSD修正



packet/md5.c
/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */
#include <string.h>
#include "md5.h"

static void MD5Init (MD5_CTX * context)
{
    context->count[0] = context->count[1] = 0;
    /* Load magic initialization constants.
    */
    context->state[0] = 0x67452301;
    context->state[1] = 0xefcdab89;
    context->state[2] = 0x98badcfe;
    context->state[3] = 0x10325476;
}
/* Constants for MD5Transform routine. */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21
/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))
/* ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
static void Encode (unsigned char *output, u_int32_t *input, unsigned int len)
{
    unsigned int i, j;

    for (i = 0, j = 0; j < len; i++, j += 4) {
        output[j] = (unsigned char)(input[i] & 0xff);
        output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
        output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
        output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
    }
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4.
 */
static void Decode (u_int32_t *output, unsigned char *input, unsigned int len)
{
    unsigned int i, j;

    for (i = 0, j = 0; j < len; i++, j += 4)
        output[i] = ((u_int32_t)input[j]) | (((u_int32_t)input[j+1]) << 8) |
                    (((u_int32_t)input[j+2]) << 16) | (((u_int32_t)input[j+3]) << 24);
}

/* MD5 basic transformation. Transforms state based on block.
 */
static void MD5Transform (u_int32_t state[4],u_char block[64])
{
        u_int32_t a = state[0], b = state[1], c = state[2], d = state[3], x[16];

    Decode (x, block, 64);

    /* Round 1 */
    FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
    FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
    FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
    FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
    FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
    FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
    FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
    FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
    FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
    FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
    FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
    FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
    FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
    FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
    FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
    FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

    /* Round 2 */
    GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
    GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
    GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
    GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
    GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
    GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
    GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
    GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
    GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
    GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
    GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */


    GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
    GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
    GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
    GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
    GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

    /* Round 3 */
    HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
    HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
    HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
    HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
    HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
    HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
    HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
    HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
    HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
    HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
    HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
    HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
    HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
    HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
    HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
    HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

    /* Round 4 */
    II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
    II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
    II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
    II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
    II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
    II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
    II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
    II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
    II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
    II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
    II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
    II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
    II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
    II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
    II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
    II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

    state[0] += a;
    state[1] += b;
    state[2] += c;
    state[3] += d;

    /* Zeroize sensitive information.

    */
    memset ((POINTER)x, 0, sizeof (x));
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
static void MD5Update (MD5_CTX *context, unsigned char *input, unsigned int inputLen)

{
    unsigned int i, index, partLen;

    /* Compute number of bytes mod 64 */
    index = (unsigned int)((context->count[0] >> 3) & 0x3F);

    /* Update number of bits */
    if ((context->count[0] += ((u_int32_t)inputLen << 3))
            < ((u_int32_t)inputLen << 3))
        context->count[1]++;
    context->count[1] += ((u_int32_t)inputLen >> 29);

    partLen = 64 - index;

    /* Transform as many times as possible.
    */
    if (inputLen >= partLen) {
        memcpy((POINTER) &context->buffer[index], (POINTER) input, partLen);
        MD5Transform (context->state, context->buffer);

        for (i = partLen; i + 63 < inputLen; i += 64)
            MD5Transform (context->state, &input[i]);

        index = 0;
    }
    else
        i = 0;

    /* Buffer remaining input */
    memcpy
    ((POINTER)&context->buffer[index], (POINTER)&input[i],
     inputLen-i);
}
static void MD5Final (unsigned char digest[16], MD5_CTX *context)
{
    unsigned char bits[8];
    unsigned int index, padLen;

    /* Save number of bits */
    Encode (bits, context->count, 8);

    /* Pad out to 56 mod 64.
    */
        u_char PADDING[64] =
        {
                0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
        };

    index = (unsigned int)((context->count[0] >> 3) & 0x3f);
    padLen = (index < 56) ? (56 - index) : (120 - index);
    MD5Update (context, PADDING, padLen);

    /* Append length (before padding) */
    MD5Update (context, bits, 8);

    /* Store state in digest */
    Encode (digest, context->state, 16);

    /* Zeroize sensitive information.
    */
    memset ((POINTER)context, 0, sizeof (*context));
}
void Computehash(u_char * src, int i,u_char*digest)
{
        MD5_CTX ct;
        MD5Init(&ct);
        MD5Update(&ct, src, i);
        MD5Final(digest, &ct);
}
